File: packages/mcp/src/cache.ts
/**
 * @settlegrid/mcp - LRU Cache for key validation
 *
 * In-memory Least-Recently-Used cache with TTL-based expiration.
 * Reduces API calls for repeated key validation within the TTL window.
 *
 * @packageDocumentation
 */

import type { KeyValidationResult } from './types'

interface CacheEntry {
  value: KeyValidationResult
  expiresAt: number
}

/**
 * LRU (Least-Recently-Used) cache with TTL expiration for key validation results.
 *
 * Used internally by the SDK to avoid redundant API calls when the same key
 * is validated multiple times within the TTL window.
 *
 * @example
 * ```typescript
 * const cache = new LRUCache(1000, 300000) // 1000 entries, 5 min TTL
 * cache.set('sg_live_abc', validationResult)
 * const cached = cache.get('sg_live_abc') // returns result or undefined
 * ```
 */
export class LRUCache {
  private readonly maxSize: number
  private readonly ttlMs: number
  private readonly cache: Map<string, CacheEntry>

  /**
   * Create a new LRU cache.
   *
   * @param maxSize - Maximum number of entries before LRU eviction (default: 1000)
   * @param ttlMs - Time-to-live in milliseconds for cache entries (default: 300000 / 5 minutes)
   */
  constructor(maxSize: number = 1000, ttlMs: number = 5 * 60 * 1000) {
    this.maxSize = maxSize
    this.ttlMs = ttlMs
    this.cache = new Map()
  }

  /** Get a cached validation result, or undefined if not found/expired */
  get(key: string): KeyValidationResult | undefined {
    const entry = this.cache.get(key)
    if (!entry) return undefined

    if (Date.now() > entry.expiresAt) {
      this.cache.delete(key)
      return undefined
    }

    // Move to end (most recently used) by re-inserting
    this.cache.delete(key)
    this.cache.set(key, entry)
    return entry.value
  }

  /** Cache a validation result */
  set(key: string, value: KeyValidationResult): void {
    // Delete first to update insertion order
    this.cache.delete(key)

    // Evict oldest if at capacity
    if (this.cache.size >= this.maxSize) {
      const oldestKey = this.cache.keys().next().value
      if (oldestKey !== undefined) {
        this.cache.delete(oldestKey)
      }
    }

    this.cache.set(key, {
      value,
      expiresAt: Date.now() + this.ttlMs,
    })
  }

  /** Invalidate a specific key */
  invalidate(key: string): void {
    this.cache.delete(key)
  }

  /** Clear the entire cache */
  clear(): void {
    this.cache.clear()
  }

  /** Get current cache size */
  get size(): number {
    return this.cache.size
  }

  /** Remove all expired entries */
  prune(): number {
    const now = Date.now()
    let pruned = 0
    for (const [key, entry] of this.cache) {
      if (now > entry.expiresAt) {
        this.cache.delete(key)
        pruned++
      }
    }
    return pruned
  }
}

File: packages/sdk-python/settlegrid/cache.py
"""LRU + TTL cache for SettleGrid key-validation results.

1:1 mirror of the TS SDK's :ts:class:`LRUCache` (``packages/mcp/src/cache.ts``).
Stores both POSITIVE and NEGATIVE results — caching ``valid=False`` lookups is
deliberate so a flood of bad keys doesn't hammer the API.

Implementation note: thin wrapper around :class:`cachetools.TTLCache`. The
TTL is enforced by ``cachetools`` itself; LRU eviction kicks in when the
cache reaches ``max_size``. ``invalidate`` and ``clear`` methods mirror the
TS surface; ``prune`` triggers an explicit expiry sweep.
"""

from __future__ import annotations

import threading
from collections.abc import Hashable
from typing import Generic, TypeVar

from cachetools import TTLCache  # type: ignore[import-untyped]

from ._types import KeyValidationResult

# Default cache parameters mirror the TS SDK exactly:
# 1000 entries, 5-minute TTL.
DEFAULT_MAX_SIZE = 1000
DEFAULT_TTL_SECONDS = 5 * 60

T = TypeVar("T", bound=Hashable)


class LRUCache(Generic[T]):
    """LRU cache with TTL expiration.

    Generic over the cached value type so the same class can serve both
    :class:`KeyValidationResult` and (in the future) other lookup results.
    The default type parameter (used by the SDK) is
    :class:`KeyValidationResult`.

    Thread-safe: ``cachetools.TTLCache`` is not thread-safe by itself, so a
    single :class:`threading.Lock` guards every public operation. The lock
    is held briefly (constant time per operation) and never crosses I/O.

    Hostile pre-checks:

    - Negative TTL / max_size raise :class:`ValueError` at construction.
    - ``get`` returns ``None`` (not raises) for missing or expired entries
      — matches the TS SDK's ``undefined`` return semantics.
    - ``invalidate`` and ``clear`` are idempotent.
    """

    def __init__(
        self,
        max_size: int = DEFAULT_MAX_SIZE,
        ttl_seconds: float = DEFAULT_TTL_SECONDS,
    ) -> None:
        if max_size < 1:
            raise ValueError(
                f"LRUCache max_size must be >= 1, got {max_size}"
            )
        if ttl_seconds <= 0:
            raise ValueError(
                f"LRUCache ttl_seconds must be > 0, got {ttl_seconds}"
            )
        self._max_size = max_size
        self._ttl_seconds = ttl_seconds
        self._cache: TTLCache[str, T] = TTLCache(
            maxsize=max_size,
            ttl=ttl_seconds,
        )
        self._lock = threading.Lock()

    def get(self, key: str) -> T | None:
        """Return the cached value, or ``None`` if missing or expired."""
        with self._lock:
            try:
                # ``TTLCache.__getitem__`` triggers expiry-on-access for
                # this single key, so an expired entry returns KeyError
                # (no manual now > expiresAt check needed).
                value: T = self._cache[key]
                return value
            except KeyError:
                return None

    def set(self, key: str, value: T) -> None:
        """Cache ``value`` under ``key``. Evicts oldest if at capacity."""
        with self._lock:
            # ``TTLCache`` handles capacity-bound LRU eviction on assignment.
            self._cache[key] = value

    def invalidate(self, key: str) -> None:
        """Remove a single key. Idempotent — missing keys are silently OK."""
        with self._lock:
            self._cache.pop(key, None)

    def clear(self) -> None:
        """Clear the entire cache."""
        with self._lock:
            self._cache.clear()

    @property
    def size(self) -> int:
        """Current number of *non-expired* entries (post-expiry sweep)."""
        with self._lock:
            # Trigger expiry of stale entries before reporting the count
            # so ``size`` matches what ``get`` would observe.
            self._cache.expire()
            return len(self._cache)

    def prune(self) -> int:
        """Force-expire stale entries; return the count removed.

        Mirrors :ts:meth:`LRUCache.prune` in the TS SDK. Useful for tests
        and for long-lived processes that want to reclaim memory between
        validation bursts.
        """
        with self._lock:
            # cachetools.TTLCache.__len__ triggers `expire()` internally,
            # which would auto-sweep before our manual sweep and report
            # zero items removed. Read the raw underlying dict instead
            # (``_Cache__data`` is the base ``Cache``'s OrderedDict, which
            # is a dict view that ignores expiry).
            raw = self._cache._Cache__data
            before = len(raw)
            self._cache.expire()
            after = len(raw)
            return before - after


# Type alias used by the SDK client + middleware so the cache type is
# unambiguous at the call site.
KeyValidationCache = LRUCache[KeyValidationResult]


__all__ = [
    "DEFAULT_MAX_SIZE",
    "DEFAULT_TTL_SECONDS",
    "KeyValidationCache",
    "LRUCache",
]
